File: UnmanagedToManaged/Engine/Transform/MarshalTypeFactory.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.
Imports System.CodeDom
Imports System.Collections.Generic
Imports PInvoke
Imports PInvoke.Parser
Imports PInvoke.Transform

Namespace Transform

    ''' <summary>
    ''' Used for creating types that help with Marshalling
    ''' </summary>
    ''' <remarks></remarks>
    Friend Module MarshalTypeFactory

#Region "PInvokePointer"

        Public Const PInvokePointerTypeName As String = "PInvokePointer"

        Public Function CreatePInvokePointerCodeTypeReference() As CodeTypeReference
            Return New CodeTypeReference(PInvokePointerTypeName)
        End Function

        Public Function CreatePInvokePointerType() As CodeTypeDeclaration
            Dim ctd As New CodeTypeDeclaration(PInvokePointerTypeName)

            ctd.Attributes = MemberAttributes.Public
            ctd.Members.Add(New CodeMemberField( _
                New CodeTypeReference(GetType(IntPtr)), _
                "m_ptr"))
            ctd.Members.Add(New CodeMemberField( _
                New CodeTypeReference(GetType(Int32)), _
                "m_size"))

            ' Add the pointer property
            Dim prop As New CodeMemberProperty
            prop.Name = "IntPtr"
            prop.HasGet = True
            prop.HasSet = False
            prop.Type = New CodeTypeReference(GetType(IntPtr))
            prop.Attributes = MemberAttributes.Public
            prop.GetStatements.Add(New CodeMethodReturnStatement( _
                New CodeVariableReferenceExpression("m_ptr")))
            ctd.Members.Add(prop)

            ' Add the size property
            prop = New CodeMemberProperty
            prop.Name = "Size"
            prop.HasGet = True
            prop.HasSet = False
            prop.Attributes = MemberAttributes.Public
            prop.Type = New CodeTypeReference(GetType(Int32))
            prop.GetStatements.Add(New CodeMethodReturnStatement( _
                New CodeVariableReferenceExpression("m_size")))

            ' Add the constructor
            Dim ctor As New CodeConstructor()
            ctor.Attributes = MemberAttributes.Public
            ctor.Parameters.Add(New CodeParameterDeclarationExpression( _
                New CodeTypeReference(GetType(IntPtr)), _
                "ptr"))
            ctor.Parameters.Add(New CodeParameterDeclarationExpression( _
                New CodeTypeReference(GetType(Int32)), _
                "size"))
            ctor.Statements.Add(New CodeAssignStatement( _
                New CodeVariableReferenceExpression("m_ptr"), _
                New CodeVariableReferenceExpression("ptr")))
            ctor.Statements.Add(New CodeAssignStatement( _
                New CodeVariableReferenceExpression("m_size"), _
                New CodeVariableReferenceExpression("size")))
            ctd.Members.Add(ctor)

            ' Add the other constructor
            ctor = New CodeConstructor()
            ctor.Attributes = MemberAttributes.Public
            ctor.Parameters.Add(New CodeParameterDeclarationExpression( _
                New CodeTypeReference(GetType(Int32)), _
                "size"))
            ctor.Statements.Add(New CodeAssignStatement( _
                New CodeVariableReferenceExpression("m_ptr"), _
                New CodeMethodInvokeExpression( _
                    New CodeTypeReferenceExpression(GetType(System.Runtime.InteropServices.Marshal)), _
                    "AllocCoTaskMem", _
                    New CodeVariableReferenceExpression("size"))))
            ctor.Statements.Add(New CodeAssignStatement( _
                New CodeVariableReferenceExpression("m_size"), _
                New CodeVariableReferenceExpression("size")))
            ctd.Members.Add(ctor)

            ' Add the free method
            Dim method As New CodeMemberMethod()
            method.Name = "Free"
            method.Attributes = MemberAttributes.Public
            method.Statements.Add(New CodeMethodInvokeExpression( _
                New CodeTypeReferenceExpression(GetType(System.Runtime.InteropServices.Marshal)), _
                "FreeCoTaskMem", _
                New CodeVariableReferenceExpression("m_ptr")))
            method.Statements.Add(New CodeAssignStatement( _
                New CodeVariableReferenceExpression("m_ptr"), _
                New CodeFieldReferenceExpression( _
                    New CodeTypeReferenceExpression(GetType(IntPtr)), _
                    "Zero")))
            method.Statements.Add(New CodeAssignStatement( _
                New CodeVariableReferenceExpression("m_size"), _
                New CodePrimitiveExpression(0)))
            ctd.Members.Add(method)

            ' Add the method to create a byte[] over the pointer
            method = New CodeMemberMethod()
            method.Name = "ToByteArray"
            method.Attributes = MemberAttributes.Public
            method.Statements.Add(New CodeVariableDeclarationStatement( _
                New CodeTypeReference(New CodeTypeReference(GetType(Byte)), 1), _
                "arr", _
                New CodeArrayCreateExpression( _
                    New CodeTypeReference(New CodeTypeReference(GetType(Byte)), 1), _
                    New CodeVariableReferenceExpression("m_size"))))
            method.Statements.Add(New CodeMethodInvokeExpression( _
                New CodeTypeReferenceExpression(GetType(System.Runtime.InteropServices.Marshal)), _
                "Copy", _
                New CodeVariableReferenceExpression("m_ptr"), _
                New CodeVariableReferenceExpression("arr"), _
                New CodePrimitiveExpression(0), _
                New CodeVariableReferenceExpression("m_size")))
            method.Statements.Add(New CodeMethodReturnStatement( _
                New CodeVariableReferenceExpression("arr")))
            method.ReturnType = New CodeTypeReference(New CodeTypeReference(GetType(Byte)), 1)
            ctd.Members.Add(method)

            AddPInvokePointerDisposeLogic(ctd)

            Return ctd
        End Function

        Private Sub AddPInvokePointerDisposeLogic(ByVal ctd As CodeTypeDeclaration)
            ctd.BaseTypes.Add(New CodeTypeReference(GetType(IDisposable)))

            Dim mem As CodeMemberMethod = New CodeMemberMethod()
            mem.Name = "Dispose"
            mem.Attributes = MemberAttributes.Public
            mem.ImplementationTypes.Add(New CodeTypeReference(GetType(IDisposable)))

            ' Just call the free method
            Dim state As New CodeExpressionStatement( _
                New CodeMethodInvokeExpression( _
                    New CodeMethodReferenceExpression(New CodeThisReferenceExpression(), "Free")))
            mem.Statements.Add(state)

            ctd.Members.Add(mem)
        End Sub

#End Region

    End Module

End Namespace

File: PInvoke.Core/Transform/MarshalTypeFactory.cs
// Copyright (c) Microsoft Corporation.  All rights reserved.
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.CodeDom;
using PInvoke;
using PInvoke.Parser;
using PInvoke.Transform;

namespace PInvoke.Transform
{

    /// <summary>
    /// Used for creating types that help with Marshalling
    /// </summary>
    /// <remarks></remarks>
    public static class MarshalTypeFactory
    {

        #region "PInvokePointer"


        public const string PInvokePointerTypeName = "PInvokePointer";
        public static CodeTypeReference CreatePInvokePointerCodeTypeReference()
        {
            return new CodeTypeReference(PInvokePointerTypeName);
        }

        public static CodeTypeDeclaration CreatePInvokePointerType()
        {
            CodeTypeDeclaration ctd = new CodeTypeDeclaration(PInvokePointerTypeName);

            ctd.Attributes = MemberAttributes.Public;
            ctd.Members.Add(new CodeMemberField(new CodeTypeReference(typeof(IntPtr)), "m_ptr"));
            ctd.Members.Add(new CodeMemberField(new CodeTypeReference(typeof(Int32)), "m_size"));

            // Add the pointer property
            CodeMemberProperty prop = new CodeMemberProperty();
            prop.Name = "IntPtr";
            prop.HasGet = true;
            prop.HasSet = false;
            prop.Type = new CodeTypeReference(typeof(IntPtr));
            prop.Attributes = MemberAttributes.Public;
            prop.GetStatements.Add(new CodeMethodReturnStatement(new CodeVariableReferenceExpression("m_ptr")));
            ctd.Members.Add(prop);

            // Add the size property
            prop = new CodeMemberProperty();
            prop.Name = "Size";
            prop.HasGet = true;
            prop.HasSet = false;
            prop.Attributes = MemberAttributes.Public;
            prop.Type = new CodeTypeReference(typeof(Int32));
            prop.GetStatements.Add(new CodeMethodReturnStatement(new CodeVariableReferenceExpression("m_size")));

            // Add the constructor
            CodeConstructor ctor = new CodeConstructor();
            ctor.Attributes = MemberAttributes.Public;
            ctor.Parameters.Add(new CodeParameterDeclarationExpression(new CodeTypeReference(typeof(IntPtr)), "ptr"));
            ctor.Parameters.Add(new CodeParameterDeclarationExpression(new CodeTypeReference(typeof(Int32)), "size"));
            ctor.Statements.Add(new CodeAssignStatement(new CodeVariableReferenceExpression("m_ptr"), new CodeVariableReferenceExpression("ptr")));
            ctor.Statements.Add(new CodeAssignStatement(new CodeVariableReferenceExpression("m_size"), new CodeVariableReferenceExpression("size")));
            ctd.Members.Add(ctor);

            // Add the other constructor
            ctor = new CodeConstructor();
            ctor.Attributes = MemberAttributes.Public;
            ctor.Parameters.Add(new CodeParameterDeclarationExpression(new CodeTypeReference(typeof(Int32)), "size"));
            ctor.Statements.Add(new CodeAssignStatement(new CodeVariableReferenceExpression("m_ptr"), new CodeMethodInvokeExpression(new CodeTypeReferenceExpression(typeof(System.Runtime.InteropServices.Marshal)), "AllocCoTaskMem", new CodeVariableReferenceExpression("size"))));
            ctor.Statements.Add(new CodeAssignStatement(new CodeVariableReferenceExpression("m_size"), new CodeVariableReferenceExpression("size")));
            ctd.Members.Add(ctor);

            // Add the free method
            CodeMemberMethod method = new CodeMemberMethod();
            method.Name = "Free";
            method.Attributes = MemberAttributes.Public;
            method.Statements.Add(new CodeMethodInvokeExpression(new CodeTypeReferenceExpression(typeof(System.Runtime.InteropServices.Marshal)), "FreeCoTaskMem", new CodeVariableReferenceExpression("m_ptr")));
            method.Statements.Add(new CodeAssignStatement(new CodeVariableReferenceExpression("m_ptr"), new CodeFieldReferenceExpression(new CodeTypeReferenceExpression(typeof(IntPtr)), "Zero")));
            method.Statements.Add(new CodeAssignStatement(new CodeVariableReferenceExpression("m_size"), new CodePrimitiveExpression(0)));
            ctd.Members.Add(method);

            // Add the method to create a byte[] over the pointer
            method = new CodeMemberMethod();
            method.Name = "ToByteArray";
            method.Attributes = MemberAttributes.Public;
            method.Statements.Add(new CodeVariableDeclarationStatement(new CodeTypeReference(new CodeTypeReference(typeof(byte)), 1), "arr", new CodeArrayCreateExpression(new CodeTypeReference(new CodeTypeReference(typeof(byte)), 1), new CodeVariableReferenceExpression("m_size"))));
            method.Statements.Add(new CodeMethodInvokeExpression(new CodeTypeReferenceExpression(typeof(System.Runtime.InteropServices.Marshal)), "Copy", new CodeVariableReferenceExpression("m_ptr"), new CodeVariableReferenceExpression("arr"), new CodePrimitiveExpression(0), new CodeVariableReferenceExpression("m_size")));
            method.Statements.Add(new CodeMethodReturnStatement(new CodeVariableReferenceExpression("arr")));
            method.ReturnType = new CodeTypeReference(new CodeTypeReference(typeof(byte)), 1);
            ctd.Members.Add(method);

            AddPInvokePointerDisposeLogic(ctd);

            return ctd;
        }

        private static void AddPInvokePointerDisposeLogic(CodeTypeDeclaration ctd)
        {
            ctd.BaseTypes.Add(new CodeTypeReference(typeof(IDisposable)));

            CodeMemberMethod mem = new CodeMemberMethod();
            mem.Name = "Dispose";
            mem.Attributes = MemberAttributes.Public;
            mem.ImplementationTypes.Add(new CodeTypeReference(typeof(IDisposable)));

            // Just call the free method
            CodeExpressionStatement state = new CodeExpressionStatement(new CodeMethodInvokeExpression(new CodeMethodReferenceExpression(new CodeThisReferenceExpression(), "Free")));
            mem.Statements.Add(state);

            ctd.Members.Add(mem);
        }

        #endregion

    }
}
